src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,562 @@
+    ///
+    /// The `incoming` and `outgoing` arguments give $B_I$ and $B_O$ respectively.
+    /// They can be either source or target ports. We currently assume that if
+    /// the source port of an outgoing boundary edge is linked to multiple
+    /// target ports, then all edges from the same source port are outgoing
+    /// boundary edges.
+    ///
+    /// More formally, the sibling subgraph defined by $B_I$ and $B_O$ is the
+    /// graph given by the connected components of the graph with edges
+    /// $E\B_I\B_O$ which contain at least one node that is either
+    ///  - the target of an incoming boundary edge, or
+    ///  - the source of an outgoing boundary edge.
+    ///
+    /// A subgraph is well-formed if for every edge in the HUGR
+    ///  - it is in $B_I$ if and only if it has a source outside of the subgraph
+    ///    and a target inside of it, and
+    ///  - it is in $B_O$ if and only if it has a source inside of the subgraph
+    ///    and a target outside of it.
+    ///
+    /// This function fails if the subgraph if it is not convex, if the nodes
+    /// do not share a common parent or if the subgraph is empty.
+    ///
+    /// The order of the boundary edges is used to determine the order of the
+    /// signature.
+    pub fn from_boundary_edges(
+        base: &'g Base,
+        incoming: impl IntoIterator<Item = (Node, Port)>,
+        outgoing: impl IntoIterator<Item = (Node, Port)>,
+    ) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        let mut checker = ConvexChecker::new(base.portgraph());
+        Self::from_boundary_edges_with_checker(base, incoming, outgoing, &mut checker)
+    }
+
+    /// Create a new sibling subgraph from some boundary edges.
+    ///
+    /// Provide a [`ConvexChecker`] instance to avoid constructing one for
+    /// faster convexity check. If you do not have one, use
+    /// [`SiblingSubgraph::from_boundary_edges`].
+    ///
+    /// Refer to [`SiblingSubgraph::from_boundary_edges`] for the full
+    /// documentation.
+    pub fn from_boundary_edges_with_checker(
+        base: &'g Base,
+        incoming: impl IntoIterator<Item = (Node, Port)>,
+        outgoing: impl IntoIterator<Item = (Node, Port)>,
+        checker: &mut ConvexChecker<&'g Base::Portgraph>,
+    ) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        let pg = base.portgraph();
+        let incoming = incoming.into_iter().flat_map(|(n, p)| match p.direction() {
+            Direction::Outgoing => base.linked_ports(n, p).collect(),
+            Direction::Incoming => vec![(n, p)],
+        });
+        let outgoing = outgoing.into_iter().flat_map(|(n, p)| match p.direction() {
+            Direction::Incoming => base.linked_ports(n, p).collect(),
+            Direction::Outgoing => vec![(n, p)],
+        });
+        let to_pg = |(n, p): (Node, Port)| pg.port_index(n.index, p.offset).expect("invalid port");
```

Comment:
this might be a place to look for state-order edges and panic(?) if you find them. Else you might comment here that we probably need to deal with such here too (?)